### ISSUE TITLE
populateDefaults not working on extensions

### ISSUE THREAD

**Issue by Firesphere** [silverstripe/silverstripe-framework#4878]
It seems, extending SiteConfig with a file and having a populate defaults breaks the build because SiteConfig is setup before File?

**Comment by jonom** [comment]
I've run in to this type of problem before and finally tried to get to the bottom of it. `populateDefaults()` shouldn't get called while the database tables are getting built, but it does. I think I've tracked down the cause but not sure the best approach for a fix.

**Background**
The `singleton()` function from Core.php is used [during the build process](https://github.com/silverstripe/silverstripe-framework/blob/3/model/DatabaseAdmin.php#L234) to create a singleton instance for each class that needs a table in the database and uses `Injector::get($className)` to do so. 

The DataObject class allows a second argument `$isSingleton` in its constructor which is used to limit the amount of initialisation done on a new instance. Notably if this is set to true it prevents `populateDefaults()` being called.

The problem is that `singleton($className)` results in an object being created with no arguments passed to the constructor. This means `$isSingleton` is assumed to be false, which leads to `populateDefaults()` being called and an SQL error thrown if required tables aren't built yet.

**Solution...?**
Think we need to fix *something* about how we approach singletons but I don't know if this is something that should be addressed in the Injector, build process, or if it's a broader architectural concern. @silverstripe/core-team any thoughts?

**Comment by tractorcow** [comment]
I think @jonom might have the right idea; Although we haven't been very careful in how we apply singleton construction to dataobjects.

We do have the Object::singleton() method, which COULD be overridden in DataObject to force supply the `true` value to the DataObject constructor's `$isSingleton` parameter. Then we could just make the schema update mechanism use `$dataClass::singleton()` instead.

On the other hand, it still feels over-engineered, but it will solve the problem in this case.

**Comment by tractorcow** [comment]
@Firesphere also if you have a custom populateDefaults, you may want to include some basic DB smoke testing to `return;` if certain required schema isn't available.

**Comment by Firesphere** [comment]
Wouldn't it be a more elegant solution, to really move the population to the "creating default records" part, after the actual build of the database?

**Comment by tractorcow** [comment]
populateDefault() is a mechanism for all dataobject instances created on the fly; It's not something you can pre-generate once and never have to set again.

**Comment by Firesphere** [comment]
I must admit, it's almost 11PM for me, the idea just came to mind ;)

We'll leave it for now :) Let's talk about it next month :) I think there's quite some improvement to be made if it's well thought-out (is that a useable word in English?)

**Comment by tractorcow** [comment]
I think we also need a better `database is ready` check for code that does bootstrapping, and could also run during dev/build.

**Comment by jonom** [comment]
Agreed - Security::database_is_ready( ) only checks that a handful of tables are ready, the ones that are required for checking privileges. This worked for https://github.com/silverstripe/silverstripe-blog/pull/140/files because we were only interested in Member but I don't think we have an equivalent check that all tables are built and ready?

[DatabaseAdmin::autoBuild()](http://api.silverstripe.org/3.2/source-class-DatabaseAdmin.html#124-137) seems to provide an implementation of this functionality although it relies on file modification times... might be nicer if we could [check the db directly](http://stackoverflow.com/questions/1525784/mysql-check-if-a-table-exists-without-throwing-an-exception) instead.

If we had this method I suppose the `$isSingleton` argument could be made redundant at least as it pertains to `populateDefaults()` as we could check that the database is ready instead of checking whether `$isSingleton` is true, and populate defaults regardless whether it's a singleton or not?

**Comment by Firesphere** [comment]
Just another random thought, I haven't looked into it. But isn't it an option to have the defaultRecords check if the database exists, and if not, put the to-create default records at the end of a queue?

A bit like Chris his demoed async PHP methods, although it would require a HUGE rewrite I guess, it would queue the actions, and if the action throws an error, requeue it to retry at a later point.

I know it's not ideal. But I'm unsure if there even is a perfect solution to this. Especially if some module has dependencies back and forward dependencies, it would become an endless queue.

I also agree with @jonom, the autoBuild should rely on the actual difference between the database and the declared records, not the modification-time. Although I think that should be a different issue.

**Comment by jonom** [comment]
requireDefaultRecords() is already run only after the database is fully built - this is the only time it happens I believe. The problem is that populateDefaults() (i.e. populate default values) is getting called needlessly while the tables are being built.

**Comment by Firesphere** [comment]
Well, I would personally go for a pragmatic resolution. One that works and takes the least time ;)

It's not something that's out of the ordinary. In this case, I think just running it after the database build would suffice just fine.
Just a thought. It shouldn't be overly complex for developers to use I'd say.

**Comment by tractorcow** [comment]
I think the problem is that, at times, SiteConfig::current_site_config(); is getting called before the DB is ready.

In the past we've just hoped that this is never invoked, but it still happens from time to time.

**Comment by jonom** [comment]
@Firesphere can you provide example of code that causes the break? Does the table building process get partway through before erroring or not even start up? If it's the former I think the cause is probably populateDefaults() getting called on SiteConfig while it's tables are being built but if it's the former @tractorcow could be on to something.

**Comment by Firesphere** [comment]
As far as I can see, Werner's Folder-Per-Page breaks on my NewsModule. At least the previous version did. 
Here's the commit which originally broke the build.
https://github.com/Firesphere/silverstripe-newsmodule/commit/7374e77d1126174785be0b36756de2c864f585b9

**Comment by wernerkrauss** [comment]
@Firesphere are you on SS3.2? My module didn't run on 3.2, thanks to travis i noticed this and hopefully fixed it.

While running tests on it I also had to de-install your news module, cause they didn't run because of this populateDefault issue...

**Comment by Firesphere** [comment]
Yes, 3.2 but also 3.1 has this issue. It's the Populate Defaults that causes it. See the commit mentioned above for a temporary fix.

**Comment by tractorcow** [comment]
I wonder if we could shiv a "treat db as incomplete" param to the command line args of sake dev/build. :P

**Comment by adamjudd** [comment]
I just ran into this for a project of mine as well. The `dev/build` runs through the page creation until it hits a model that uses SiteConfig within it's `populateDefaults()` and errors at that point, preventing it from continuing.

**Comment by tractorcow** [comment]
This might still break if you add an extension that adds a column to SiteConfig; The table will exist, but attempts to query it will probably fail.

**Comment by adamjudd** [comment]
@tractorcow, great point - and that is actually how I originally ran into this issue, but my workaround worked when re-building the database for this project, so siteconfig didn't exist then.

**Comment by jonom** [comment]
I've run in to this problem again today. @tractorcow what do you think of this solution? https://github.com/jonom/silverstripe-framework/commit/bc8d5be536cfa2a883feb3b02598e62fcfe4f9fa

**Comment by tractorcow** [comment]
I think the idea is in the right direction, but we probably need a more robust "table can be queried" check.

The problem is expense. I would go so far as to say that we should probably just allow errors to blow the site up when NOT performing a dev/build. However, actually do this check when tableBuildInProgress() is true, as we can't realistically query the whole DB schema every request.

It also feels risky to assume that the schema is invalid just because tableBuildInProgress() is set. Sometimes (e.g. when checking if the member is allowed to dev/build) we need to query these tables.

`Security::db_is_ready` was a prior attempt at this. I'd look at replacing this functionality at the same time into a consistent API.

Maybe it would be on `DataObject::schema_ready()`, which you could call on various dataobjects?

**Comment by jonom** [comment]
@tractorcow I agree with everything you're saying, I just don't have the resources to come up with a more robust solution at the moment. However, the root of this particular issue is a bug in singleton production and I think I may have a fix for that, which in turn solves this issue. I can never seem to get all tests to run locally so I've opened a PR to let them run: https://github.com/silverstripe/silverstripe-framework/pull/5151

**Comment by tractorcow** [comment]
Oh, I realise this... and I'm in the same situation, or I'd have love to address this myself.

I think your fix is fine, and I've left a comment on it.

**Comment by jonom** [comment]
This is still broken in 3.6 so I opened a new PR at https://github.com/silverstripe/silverstripe-framework/pull/7850

**Comment by jonom** [comment]
@dhensby should this be re-opened as a separate fix is needed for 4?

**Comment by dhensby** [comment]
maybe - is it actually the same issue, though?

**Comment by jonom** [comment]
I think so: (emphasis mine)

> extending SiteConfig with a file **and having a populate defaults** breaks the build because SiteConfig is setup before File?

The issue is that `populateDefaults()` is being executed on singletons during the 'Creating database tables' step of the build. `populateDefaults()` should be called during a build, but not until the 'Creating database records' step, when the tables have been built and we're working with real/non-singleton/prototype (I don't know the right term) objects. So it shouldn't matter if SiteConfig is setup before File, it's only an issue because `populateDefaults()` is running when it shouldn't.

**Comment by tractorcow** [comment]
Simpler solution:

when in a DatabaseSchema::schemaUpdate() code block, set a `ready` flag on the database to false. If you absolutely need to invoke DB queries during populadeDefaults(), you should query this prior to execution.

Putting this temp state within the callback for schemaUpdate() ensures a db is only forced to ready=false during known "unready" states, without having to actually query the DB schema to see if it's dirty.

**Comment by dhensby** [comment]
> We deprecate Security::database_is_ready() and request users use DB::get_conn()->isReady() instead.

Let's not go creating new DB APIs just to replace them in 5.x ;)

**Comment by tractorcow** [comment]
well that depends what you want in 5.

**Comment by jinjie** [comment]
Is this fixed or any work around? I have a DataExtension on SiteConfig but doesn't seem to populate the defaults.

I have tried `private static $defaults = [...]` and `populateDefaults()` nothing works.

**Comment by sminnee** [comment]
The underlying issue is a duplicate of #1575.

**Comment by KavindaHarshana** [comment]
you can use onAfterWrite:
eg: 
public function onAfterWrite()
    {
        parent::onAfterWrite();
        if(!$this->owner->Country) {
            $this->owner->Country = 'gb';
            $this->owner->write();
        }
    }
